File: src/test/java/station/DataStationTest.java
package station;

import junit.framework.TestCase;
import secret.Secret;
import secret.SecretPart;

import java.math.BigInteger;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Random;

import static secret.Secret.generateSecret;
import static util.Util.matrixDiagonalMultiplication;

public class DataStationTest extends TestCase {
    private final int POPUlATION = 10;

    public void test2PartyCalculation() {
        BigInteger[][] A_data = createData(POPUlATION);
        DataStation stationA = new DataStation("1", A_data);

        BigInteger[][] B_data = createData(POPUlATION);
        DataStation stationB = new DataStation("2", B_data);

        Secret secret = generateSecret(Arrays.asList(stationA, stationB));

        stationA.setLocalSecret(secret.getParts()[0]);
        stationB.setLocalSecret(secret.getParts()[1]);

        BigInteger ra = secret.getParts()[0].getR();
        BigInteger rb = secret.getParts()[1].getR();
        BigInteger[][] B_hat = stationB.getObfuscated();
        BigInteger[][] A_hat = stationA.getObfuscated();
        BigInteger[][] Ra = secret.getParts()[0].getMatrix();

        // manually run the calculation: A_hat * B + rb - B_hat * Ra + ra to determine expected result
        BigInteger expectedresult =
                matrixDiagonalMultiplication(Arrays.asList(A_hat, B_data), POPUlATION).add(rb)
                        .subtract(matrixDiagonalMultiplication(Arrays.asList(B_hat, Ra), POPUlATION)).add(ra);

        // run 2-party calculation
        List<BigInteger[][]> list = new ArrayList<>();
        list.add(stationA.getObfuscated());
        BigInteger partial = stationB.localCalculationFirstParty(list);
        List<BigInteger[][]> list2 = new ArrayList<>();
        list2.add(stationB.getObfuscated());
        partial = stationA.localCalculationNthParty(list2, partial);
        BigInteger result = stationB.removeV2(partial);

        assertEquals(expectedresult, result);
    }

    public void testGetObfuscated() {
        BigInteger[][] data = createData(POPUlATION);

        DataStation station = new DataStation("1", data);
        SecretPart secret = generateSecret(Arrays.asList(station)).getParts()[0];

        station.setLocalSecret(secret);

        BigInteger[][] obfuscated = station.getObfuscated();
        for (int i = 0; i < POPUlATION; i++) {
            assertEquals(data[i][i].add(secret.getMatrix()[i][i]), obfuscated[i][i]);
        }

    }

    public static DataStation createStation(String id, int population) {
        return new DataStation(id, createData(population));
    }

    private static BigInteger[][] createData(int population) {
        BigInteger[][] data = new BigInteger[population][population];

        BigInteger[][] secretMatrix = new BigInteger[population][population];
        Random random = new Random();
        for (int i = 0; i < population; i++) {
            for (int j = 0; j < population; j++) {
                data[i][j] = BigInteger.valueOf(random.nextInt(2));
            }
        }
        return data;
    }

}

File: python/test/station/DataStation_test.py

import unittest

import numpy

from station import DataStation
from secret import generateSecret
from util.Util import matrixDiagonalMultiplication

POPULATION = 4


class DataStationTest(unittest.TestCase):
    def test_2PartyCalculation(self):
        A_data = createData(POPULATION)
        stationA = DataStation.DataStation("1", A_data)
        B_data = createData(POPULATION)
        stationB = DataStation.DataStation("2", B_data)

        secret = generateSecret([stationA, stationB])

        stationA.setLocalSecret(secret.getParts()[0])
        stationB.setLocalSecret(secret.getParts()[1])

        ra = secret.getParts()[0].getR()
        rb = secret.getParts()[1].getR()

        B_hat = stationB.getObfuscated()
        A_hat = stationA.getObfuscated()

        Ra = secret.getParts()[0].getMatrix()

        # manually run the calculation: A_hat * B + rb - B_hat * Ra + ra to determine expected result
        expectedresult = matrixDiagonalMultiplication([A_hat, B_data]) + (rb) - (
            matrixDiagonalMultiplication([B_hat, Ra])) + (ra)

        # run 2 - party calculation
        list = []
        list.append(stationA.getObfuscated())
        partial = stationB.localCalculationFirstParty(list)
        list2 = []
        list2.append(stationB.getObfuscated())
        partial = partial + stationA.localCalculationNthParty(list2)
        result = stationB.removeV2(partial)

        self.assertEquals(expectedresult, result)

    def test_GetObfuscated(self):
        data = createData(POPULATION)

        station = DataStation.DataStation("1", data)
        secret = generateSecret([station]).getParts()[0]

        station.setLocalSecret(secret)

        obfuscated = station.getObfuscated()
        for i in range(0, POPULATION):
            self.assertEquals(data[i][i] + secret.getMatrix()[i][i], obfuscated[i][i])


def createStation(id: str, population: int):
    return DataStation.DataStation(str(id), createData(population))


def createData(population: int):
    data = numpy.zeros((population, population))
    for i in range(0, population):
        data[i][i] = 1 #random.randint(0, 1)
    return data
